File: Eyes/ImageTester/src/lib/java/com/applitools/ImageTester/TestObjects/ImageStep.java
package com.applitools.ImageTester.TestObjects;

import com.applitools.ImageTester.Interfaces.ITestable;
import com.applitools.ImageTester.Patterns;
import com.applitools.eyes.images.Eyes;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.util.Collection;
import java.util.regex.Pattern;

public class ImageStep extends TestUnit {
    private static final Pattern pattern = Patterns.IMAGE;
    private BufferedImage img_;

    public ImageStep(File file) {
        super(file);
    }

    public void run(Eyes eyes) throws IOException {
        try {
            eyes.checkImage(getImage(), name());
        } catch (IOException e) {
            System.out.printf("Failed to process image file: %s \n Reason: %s \n",
                    file_,
                    e.getMessage());
            throw e;
        }
    }

    public BufferedImage getImage() throws IOException {
        if (img_ == null) {
            img_ = ImageIO.read(file_);
        }
        return img_;
    }

    @Override
    public String name() {
        return super.name().replaceAll(Patterns.IMAGE_EXT, "");
    }

    public static boolean supports(File file) {
        return pattern.matcher(file.getName()).matches();
    }

    public boolean hasRegionFile() {
        return RegionStep.supports(this);
    }

    public Collection<ITestable> getRegions() throws IOException {
        return RegionStep.getSteps(this);
    }

    public void dispose() {
        if (img_ != null) img_ = null;
    }
}

File: Eyes/ImageTester/src/csharp/lib/TestObjects/ImageStep.cs
using System;
using System.Drawing;
using System.IO;
using System.Linq;
using Applitools.Images;

namespace Applitools.ImageTester.TestObjects
{
    public class ImageStep : TestUnit
    {
        private Bitmap img_;

        public static string[] SupportedExtensionsWithSeparator { get; } = new[] { ".jpg", ".png",".gif", ".bmp" };

        public ImageStep(FileStream file) : base(file)
        {
        }

        public override void Run(Eyes eyes)
        {
            try
            {
                eyes.CheckImage(GetImage(), Name());
            }
            catch (IOException e)
            {
                Console.WriteLine("Failed to process image file: {0} \\n Reason: {1} \\n", file.Name, e.Message);
                throw;
            }
        }

        public virtual Bitmap GetImage()
        {
            if (img_ == null)
            {
                img_ = new Bitmap(file);
            }

            return img_;
        }

        public override string Name()
        {
            return Path.GetFileNameWithoutExtension(base.Name());
        }

        public static bool Supports(FileStream file)
        {
            return SupportedExtensionsWithSeparator
                .AsParallel()
                .Any(e => file.Name.EndsWith(e));
        }

        public override void Dispose()
        {
            img_?.Dispose();
        }
    }
}